SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoopInvariantCodeMotion.swift

```diff
@@ -0,0 +1,1076 @@
+        guard let firstInst = succesor.instructions.first else {
+          continue
+        }
+        
+        let builder = Builder(before: firstInst, context)
+        
+        let ownership: StoreInst.StoreOwnership = firstInst.parentFunction.hasOwnership ? .trivial : .unqualified
+        builder.createStore(
+          source: ssaUpdater.getValue(inMiddleOf: succesor),
+          destination: initialAddr,
+          ownership: ownership
+        )
+        changed = true
+      }
+    }
+    
+    // In case the value is only stored but never loaded in the loop.
+    if initialLoad.uses.isEmpty {
+      context.erase(instruction: initialLoad)
+    }
+    
+    return changed
+  }
+}
+
+private extension Instruction {
+  var hasOwnershipOperandsOrResults: Bool {
+    guard parentFunction.hasOwnership else { return false }
+
+    return results.contains(where: { $0.ownership != .none })
+      || operands.contains(where: { $0.value.ownership != .none })
+  }
+  
+  /// Returns `true` if this instruction follows the default hoisting heuristic which means it
+  /// is not a terminator, allocation or deallocation and either a hoistable array semantics call or doesn't have memory effects.
+  func canBeHoisted(outOf loop: Loop, _ context: FunctionPassContext) -> Bool {
+    switch self {
+    case is TermInst, is Allocation, is Deallocation:
+      return false
+    case is ApplyInst:
+      switch arraySemanticsCallKind {
+      case .getCount, .getCapacity:
+        if canHoistArraySemanticsCall(to: loop.preheader!.terminator, context) {
+          return true
+        }
+      case .arrayPropsIsNativeTypeChecked:
+        return false
+      default:
+        break
+      }
+    default:
+      break
+    }
+
+    if memoryEffects == .noEffects {
+      return true
+    }
+
+    return false
+  }
+  
+  func hoist(outOf loop: Loop, _ context: FunctionPassContext) -> Bool {
+    guard operands.allSatisfy({ !loop.contains(block: $0.value.parentBlock) }) else {
+      return false
+    }
+
+    let terminator = loop.preheader!.terminator
+    if canHoistArraySemanticsCall(to: terminator, context) {
+      hoistArraySemanticsCall(before: terminator, context)
+    } else {
+      move(before: terminator, context)
+    }
+    
+    if let singleValueInst = self as? SingleValueInstruction,
+       !(self is BeginAccessInst),
+       let identicalInst = (loop.preheader!.instructions.first { otherInst in
+         return singleValueInst != otherInst && singleValueInst.isIdenticalTo(otherInst)
+    }) {
+      guard let identicalSingleValueInst = identicalInst as? SingleValueInstruction else {
+        return true
+      }
+      
+      singleValueInst.replace(with: identicalSingleValueInst, context)
+    }
+
+    return true
+  }
+  
+  func sink(outOf loop: Loop, _ context: FunctionPassContext) -> Bool {
+    let exitBlocks = loop.exitBlocks
+    var changed = false
+
+    for exitBlock in exitBlocks {
```

Comment:
It makes sense to add an assert here which checks that the exiting edge is not critical:
```
  assert(exitBlock.hasSinglePredecessor, "exiting edge should not be critical)
```